### DISCUSSION TITLE
In model `PageContent`, `created_by` and `changed_by` shall be normalized

### DISCUSSION THREAD

**Discussion by jrief** [django-cms/django-cms#7855] [Polls]
This is a big no-no in our database design of models: In  model `PageContent`, fields `created_by` and `changed_by` contain the full name of the user instead of their foreign key. This strongly violates the [1st normalization rule according to Edgar F. Codd](https://en.wikipedia.org/wiki/Codd%27s_12_rules) and should be out of debate.

Can anybody explain why this happened?

**Comment by marksweb** [comment]
I would suspect this is to handle users being deleted so that the name of the person persists.

**Reply by fsbraun** [reply]
If someone needed to keep information on former users would the Django way not be to deactivate the user? 
I am thinking of a field with `on_delete=models.SET_NULL`. User objects are also stored by versioning and potentially other apps.

**Reply by marksweb** [reply]
Yes historically that's how I've handled things.

**Comment by jrief** [comment]
I agree with @fsbraun that users shall be deactivated. If they want to be removed from the database (GDPR compliance), we might add a function to hash their name, email, etc. on some way. But users shall not be removed. I would even argue to use `on_delete=models.PROTECT` on the foreign keys of CMS models pointing to User.

**Reply by fsbraun** [reply]
That's an interesting take. I can imagine making this configurable and being reused throughout the CMS (esp. within versioning). Options would be `PROTECT` and `SET_NULL` and ... ? I can imagine `PROTECT` being the default giving max. accountability.

versioning has a similar setting to prevent deleting historical versions.

**Reply by jrief** [reply]
`SET_NULL` has the problem that such a `ForeigKey` then must be nullable and on each lookup we must add code to check for `None`. Since imo users shall *never* be removed, this makes the code more brittle to errors.

**Reply by sepi** [reply]
I'd be ok with not allowing to remove users as a structure but then you would definitely need to roll out an action to scramble the personal data in the same release.

The only reason I see to store the users' names instead of references would be to have a better fidelity of the audit trail. A user changing their name to obscure something would still be recorded as it was at the time. It's pretty contrived I must agree.
